### PR TITLE
psf_photometry with masks, output as astropy.table.Table

### PR DESCRIPTION
**Pull request by hamogu**
I started to add more features to the psf photometry (because I need it for a project). Unfortuntely, I did not finish it before I go on vacation, so I decided to put this half-working branch out there.

This is my plan for this PR:
- [x] psf photometry on the edge of the image
- [x] psf photometry with masks (particularly to make useful fits with nans in the aperture)
- [x] input and output with astropy tables
- [x] apply the same changes to PRF fitting

**EDIT: See updates task list further down.**

So far, all development is in `GaussianPSF`, but the same applies to other PSFs and to PRFs, too. I think I can do PRFs in this PR, but later we will need some refactoring to make this work with arbitrary PSF models.

Comments on the API changes in here are welcome, other help is appriciated, too.

### CODE REVIEW

**Comment by hamogu** [comment]
It would help if we could resolve https://github.com/astropy/imageutils/issues/21
since that can (potentially) simplify how I deal with sources on the edge of the field.

**Comment by hamogu** [comment]
I suspect that astropy/#3333 will not make it into 1.0, but I really want to use that here. Can I do that or do we need to support astropy 1.0?

**Comment by cdeil** [comment]
@hamogu At the moment we support Astropy 0.4 in photutils, after the 1.0 release we can discuss if we drop 0.4 support. You can put some version of what you need in `photutils/extern` to make it work for now, and then at some point in the future we switch to the version in Astropy.

**Comment by larrybradley** [comment]
@hamogu, are you still working (or plan to work) on this PR?

**Comment by hamogu** [comment]
@larrybradley : I started that when I worked on a project that needed PSF photometry. I have not worked on that project since 2014, but now I'm back on it. (Hence my recent PRs to photutils and ccdproc). I will get this far enough to do the photometry I need in the next few weeks and then wrap it up to make this PR mergable. 
It might not fill all the checkboxes I put in originally (e.g. I might not touch the PRF fitting), but it will help to some degree. Remaining issues can then be kept as a new issue.

**Comment by larrybradley** [comment]
Thanks, @hamogu!

**Comment by eteq** [comment]
@hamogu - in an odd quirk of timing I am re-examining this stuff too (along with @larrybradley and some others at STScI).  I'm imagining a more thorough re-vamping of some of this interface to make it somewhat more consistent with the way `modeling` works.  Of particular importance is the removal of the `fit` method to live either in the driver `aperture_photometry` function or in customized `Fitter`s that are more transparent and extendable for other kinds of PSFs.

I *think* a lot of that is orthogonal to what you're saying here though, so that's promising... Of particular interest to me right now is your third bullet, the `Table` input/output.  Have you done that already on your local version?  If so, maybe you can do that in an independent PR so that it can be dealt with separately? (Unless your changes make that too intermixed...)

**Comment by hamogu** [comment]
I an really odd quirk of timing I just started working on this again yesterday afternoon (about 3 hours before this message) and decided that it's not worth fixing this in the old way the models are defined. Also, there are far too many issues that are related in some way (#254, #189, #144, #63, #49, #48, #43), most of which can be addressed easily the migrating to the new models interface.
So, I just started out fresh to migrate the models to the new modelling interface, but I did not get very far because I had to go home early.

@eteq Maybe we can have an impromtu 10 min telecon / phone call today and see how we divy that up without dublication?
I have a seminar from 11am  - 1 pm - before or after would work.

**Comment by eteq** [comment]
@hamogu - sounds good, I sent you an email with availability (I *think* I sent it to the right address... Let me know if not.)

**Comment by hamogu** [comment]
@eteq : I added my stuff and replaced the old branch here with what you and I've come up with now.
I added two new tests to check on the mask thing, those can also serve as examples right now until we've written proper docs.

- [ ] make old tests work again (new input format for photometry)
- [ ] add tests for what you added
- [ ] write docs.

I'll continue to start ticking off those tasks as long as I'm in my office today. Please have a look and see what you think, if you want me to finish this off or which parts you want to leave to me!

**Comment by hamogu** [comment]
@eteq: Tests pass at least locally (Travis is still running). The big thing that's still missing is docs; some smaller things are tests for the store_fit_info option and some small things like that.
I think that I could convert your example notebook into proper narrative docs pretty quickly, but I don't really want to do that until we agree that the interface is close to done. 

Going forward, I suggest that we talk again today and decide what needs to be done in this PR. I think we both want to use this and build other things on top of it, so I would be totally OK with merging it fast and adding other items on the check list above as separate PRs.

File: photutils/psf.py [hunk 1]
```diff
@@ -107,94 +118,190 @@ def evaluate(self, x, y, amplitude, x_0, y_0):
...
+                sub_prf_indices = np.all(positions_subpixel_indices == [j, i],
+                                         axis=1)
+                positions_sub_prfs = positions[sub_prf_indices]
+                for k, position in enumerate(positions_sub_prfs):
+                    x, y = position
+                    extracted_prf = extract_array(data_internal, (size, size),
+                                                  (y, x))
+                    # Check shape to exclude incomplete PRFs at the boundaries
+                    # of the image
+                    if (extracted_prf.shape == (size, size) and
```
Comment (hamogu) [inline photutils/psf.py]: Should we really exclude PSFs that are only partially contained in the image? They still contribute useful information and could be treated the same way as a source in the image that has nans in it (with a mask). 
We can leave that for another PR, though.

File: photutils/psf.py [hunk 2]
```diff
@@ -107,94 +118,190 @@ def evaluate(self, x, y, amplitude, x_0, y_0):
...
+                            np.ma.sum(extracted_prf) != 0):
+                        # Replace NaN values by mirrored value, with respect
+                        # to the prf's center
```
Comment (hamogu) [inline photutils/psf.py]: I have to admit I never understood the use case for this. Assuming a symmetric PRF seems to be a special case. Do we still need that now that we have the option to use a mask and simply ignore the value? (I have ot admit that I don't know match about empirical PRFs - I usually use analytical functions in fitting myself.

File: photutils/tests/test_psfs.py [hunk 3]
```diff
@@ -19,20 +19,108 @@
...
+
+    # this is the analytic value needed to get a total flux of 1
+    mof.amplitude = (mof.alpha-1)/(np.pi*mof.gamma**2)
+
+    # first make sure it really is normalized
+    assert (1 - integrate.dblquad(mof, -10, 10, lambda x: -10, lambda x: 10)[0]) < 1e-6
+
+    # now create an "image" of the PSF
+    xg, yg = np.meshgrid(*([np.linspace(-2, 2, 100)]*2))
+    img = mof.render(coords=(xg, yg))
```
Comment (hamogu) [inline photutils/tests/test_psfs.py]: The render functionality is new in astropy 1.1. Either we simply skip these tests for all tests running the 1.0 series or we backport that, too, or we just bumb up the minimum required version (and get rid of the other compatibility stuff, too.)

Personally, I'm in favour of requiring astropy >=1.1.
I don't know when the next photutils release is planned, but until then only users of the dev version will feel the impact (and those users are probably more likely to have a current astropy anyway). When the next release comes, astropy 1.1. is hopefully already established. 
After all, the work on astropy modelling is done to ensure that we can use it - and we can only do that if we are willing to use a current version.

File: photutils/psf.py (hunk 1 above)
Comment (eteq) [inline photutils/psf.py]: I agree we should make this change, but in a separate PR. I *think* the original motivation was just "this is easier".

*But* to be clear I didn't add this - this section was just copied-and-pasted from the original implementation - the change here is to move it from being a function in the base namespace to being a classmethod.  I did that because the function's whole purpose was to create a new `DiscretePRF` and that seemed more appropriate as a class method given how closely tied it is to the `DiscretePRF` class.

File: photutils/psf.py (hunk 2 above)
Comment (eteq) [inline photutils/psf.py]: I think this is a bit of a kludge to get around the problem that `NaN`s don't play well with a lot of arithmetic.  PSFs are generally at least *somewhat* symmetric, especially for space missions.  And the use case for this class is undersampled data, so just picking the mean of the surrounding points is probably also insufficient. That said, other kinds of interpolation make as much sense, this is definitely a special case.

(As above, though, this isn't my addition - I just moved it to a different place)

File: photutils/tests/test_psfs.py (hunk 3 above)
Comment (eteq) [inline photutils/tests/test_psfs.py]: I think we want to try to continue supporting the current LTS. Even if the PSF stuff is rather experiemntal, there are plenty of "real users" on the aperture and star-finder side of things.  Forcing them to upgrade to 1.1 might be more of an issue.

But this particular use of `render` is really trivial and I now realize unnecessary.  I just checked now and it turns out ``mof(xg,yg)`` works so we can just change to that.

**Comment by eteq** [comment]
@hamogu - Are you working on this now to try to address the failing tests?  The LTS failures I think can be fixed with a trivial change I just PRed to you (hamogu/photutils#2), but I haven't had a chance to look at the others.

**Comment by hamogu** [comment]
I just merged your PR to my branch, so we can run Tavis again. I think most of the failures will go away. I'll look a the others.

**Comment by hamogu** [comment]
@eteq : Sphinx fails because your new extension requires Jupyter or IPython which is not a required dependency for photutils.
Can I suggest that, instead of rendering the notebook through a sphinx extension, we just include the generated rst file directly?

If that's all that's left, I'm willing to spend my afternoon writing proper narrative docs.

And then there us a numpy 1.6 failure. I suggest we just xfail that. Who's using numpy 1.6 nowadays?

(Will wait for Travis to run through and see if I missed anything in the list above before I do those changes.)

**Comment by larrybradley** [comment]
I'd be in favor of dropping `numpy 1.6` support.  @eteq Why does `astropy` still support it?

**Comment by hamogu** [comment]
I don't understand why the coverage in `aperture_funcs.py` dropped so much. We did not tough that file and I don't think we touched any of the tests that cover it.

**Comment by hamogu** [comment]
@eteq: Your opinion on the docs? 
I suggest we remove your extension again and just add external links to the notebook. #43 is still open which says "write PSF photometry docs". I can do that very soon, based on the notebooks, but we can merge that in a separate PR.

**Comment by larrybradley** [comment]
@hamogu  The `aperture_func` coverage dropped because of a PR I merged earlier this week (the author was unable to add tests for 6 weeks).  I plan to add the missing tests.  I'm not sure why that coverage loss is showing up here, but perhaps you should rebase.

**Comment by hamogu** [comment]
@larrybradley If we know where it's coming from I don't think a rebase is necessary. I just wanted to make sure we did not screw that up in this PR. 
Let's decide on the docs and drip numpy 1.6 and then we are good to go.

**Comment by larrybradley** [comment]
The original PSF/PRF notebook is in the https://github.com/astropy/photutils-datasets repo
(https://github.com/astropy/photutils-datasets/blob/master/notebooks/PSFPhotometrySpitzer.ipynb).
I think that old notebook should be replaced with the new version (which we could then link to from the new PSF docs).

**Comment by eteq** [comment]
@hamogu @larrybradley - I'm fine with just dropping numpy 1.6 support here. We were discussing that at one point in Astropy but I think just never finished the discussion.

**Comment by eteq** [comment]
Re: the notebooks: I actually think it would be better to try to get this to work in the repo - easy to fix on travis or RTD by just adding a jupyter dependency (I *think*).  That creates something like "poor-man's doctests", because building the docs requires that the notebooks work.  And I think it's better to keep notebooks matched to the code in this way.  I could accept just putting in the ``.rst`` for now, though, and a later PR can do better.  The main downside is that that explodes the repo due to large image files.

Leaving the `PSFPhotometrySpitzer.ipynb` in there was a mistake on my part, though - I meant to adapt that to do a better job of downloading the data with the caching machinery but never did that.  I can fix it if we want to leave them in.

**Comment by hamogu** [comment]
#327 to drop numpy 1.6 (better to do that in a clean PR and rebase this on it later)

**Comment by eteq** [comment]
@hamogu - after a bit more thought, I'm not sure it will be so easy as I said here... RTD *might* have an issue installing pandoc (which is needed for the nb->RST machinery in nbconvert).  I think the best thing is to just merge this and see, though.  If it turns out we have to do what you say (generate the RST and check that in), then we can do that in a second PR, but it's often tough to diagnose RTD problems anywhere other than the "real" branch...

**Comment by eteq** [comment]
Looks like the one remaining failure is in the spitzer notebook, and I can reproduce that locally.  Will try to fix that now, but if I can't do it by the end of today, shall we just re-name the extension so that it doesn't get run, and pull the trigger?

**Comment by hamogu** [comment]
I introduced more failures during the rebase. I'm trying to diagnose that now. I'll report back in 5-10 min.

**Comment by hamogu** [comment]
Oh, apparently I was looking at and old Travis. Agreed. Fix today or shoot.

**Comment by eteq** [comment]
Alright, I'm going to go ahead and merge this with a partial but not fully-functional, renamed version of the spitzer notebook.  Will fix in the near future.  I'll also go ahead and add some relevant issues.

**Comment by larrybradley** [comment]
Many thanks, @hamogu and @eteq!

**Comment by bsipocz** [comment]
@hamogu - This rather big PR doesn't have any trace in the changelog.
